### ISSUE TITLE
CI failure with GTSAM related to boost headers

### ISSUE THREAD

**Issue by prashanthr05** [dic-iit/kindyn-vio#19]
While trying to setup a CI with GTSAM as dependency and trying to build it from source, GTSAM fails to compile due to boost dependencies.
See the list of failures in,

https://github.com/dic-iit/kindyn-vio/actions?query=is%3Afailure+branch%3Afeature%2Fsetup-workflow

I have tried different combinations of boost versions with GTSAM to try and fix the compilation. 

Looks like Github Actions stopped supporting boost? https://github.com/actions/boost-versions
and GTSAM devs have fixed boost CI installation using custom scripts. https://github.com/borglab/gtsam/pull/720/files

For windows, they started to use https://github.com/MarkusJx/install-boost (can be configured for both windows and ubuntu but not macOS). 

I tried using the same action for both Ubuntu and Windows, in vain, unable to reach a fix.

**Comment by prashanthr05** [comment]
Looks like, the boost installed using the custom action is not being used in Ubuntu. Instead, a pre-existing boost of version 1.74 is used. See [here](https://github.com/dic-iit/kindyn-vio/runs/3154493744?check_suite_focus=true). Apparently v1.67.0 is used in [GTSAM CI](https://github.com/borglab/gtsam/blob/5bb69f72237e1a9f6ab8e46c187ebf47d06fd998/.github/workflows/build-linux.yml#L15).

**Comment by prashanthr05** [comment]
Looks like GTSAM installation passed in macOS [here](https://github.com/dic-iit/kindyn-vio/runs/3150487279?check_suite_focus=true)

**Comment by prashanthr05** [comment]
macOS CI is in place. Windows and Linux, not yet.

**Comment by prashanthr05** [comment]
cc @traversaro would you have any suggestions?

**Comment by prashanthr05** [comment]
Looks like both Windows and Linux CI's GTSAM is not getting the boost from the configured cmake variable `Boost_DIR`.

**Comment by traversaro** [comment]
On Windows, why you don't just install boost via conda?

**Comment by prashanthr05** [comment]
> On Windows/conda, why you don't just install boost via conda?

I had tried that [here](https://github.com/dic-iit/kindyn-vio/actions/runs/1062376361/workflow). It didn't seem to work.

**Comment by prashanthr05** [comment]
> > On Windows/conda, why you don't just install boost via conda?
> 
> I had tried that [here](https://github.com/dic-iit/kindyn-vio/actions/runs/1062376361/workflow). It didn't seem to work.

AH, you menat via conda channel? instead of using robotology channel? I shall try that.

**Comment by prashanthr05** [comment]
> Both Windows and Linux CI's GTSAM is not getting the boost from the configured cmake variable `Boost_DIR`.

Replaced `Boost_DIR` with `BOOST_ROOT`, `BOOST_INCLUDEDIR` and `BOOST_LIBRARYDIR` similar to https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/.github/workflows/build-windows.yml#L81

However,  manually configured cmake variables are not being used, in Ubuntu CI.
https://github.com/dic-iit/kindyn-vio/runs/3159717172?check_suite_focus=true#step:8:139

**Comment by traversaro** [comment]
> something like this?

You just need to add the `boost-cpp`  package to all the packages that are already installed via `mamba` , this comes (like `pcl` , `libmatio` , etc etc) from the conda-forge channel but you can avoid to explicitly specify it as it is set as default channel.

**Comment by prashanthr05** [comment]
At the moment, I am unable to reach a solution for both Windows and Ubuntu, and I prefer to handle this later. 

One CI is better than no CI. Currently MacOS CI is working. We can merge the branch with macOS CI to master.

**Comment by prashanthr05** [comment]
@traversaro For ubuntu, I found a packaging for gtsam. Do you think we could use this directly? 
https://anaconda.org/s_pillai/gtsam

**Comment by traversaro** [comment]
> @traversaro For ubuntu, I found a packaging for gtsam. Do you think we could use this directly?
> https://anaconda.org/s_pillai/gtsam

No, I strongly suggest not to use any conda channel outside of `conda-forge`,  `robotology` or `robostack`. The chances that they are doing stuff correctly unless the library has no dependency is quite low.

**Comment by prashanthr05** [comment]
> > @traversaro For ubuntu, I found a packaging for gtsam. Do you think we could use this directly?
> > https://anaconda.org/s_pillai/gtsam
> 
> No, I strongly suggest not to use any conda channel outside of `conda-forge`, `robotology` or `robostack`, or at least double check before using it. The chances that they are doing stuff correctly unless the library has no dependency is quite low. For gtsam, a packaging effort is ongoing in [conda-forge/staged-recipes#14339 (comment)](https://github.com/conda-forge/staged-recipes/pull/14339#issuecomment-877529408), but I think for the moment it may be blocked for several reasons. For the time being compiling gtsam from source seems to be the best options.

Noted!!

**Comment by prashanthr05** [comment]
Thanks to @traversaro, Ubuntu CI is functional now. See #22.
